// browser-metro/dist/bundler.js
import { Resolver } from "./resolver.js";
import { buildCombinedSourceMap, countNewlines, inlineSourceMap, shiftSourceMapOrigLines, } from "./source-map.js";
import { formatTransformError } from "./transform-error.js";
import { findRequires, rewriteRequires, lowerDynamicImports, buildBundlePreamble, parseExternalsFromBody, hashDeps, parseDepBundle } from "./utils.js";
export class Bundler {
    constructor(fs, config) {
        this.prefetchedPackages = {};
        this.fs = fs;
        this.config = config;
        const paths = Bundler.readTsconfigPaths(fs);
        this.resolver = new Resolver(fs, { ...config.resolver, ...(paths && { paths }) });
        this.plugins = config.plugins ?? [];
    }
    /** Prefetch all dependencies in a single batch request */
    async prefetchDependencies() {
        const versions = this.getPackageVersions();
        if (Object.keys(versions).length === 0)
            return;
        // Remove aliased and shimmed packages - they're handled client-side
        const aliases = this.getModuleAliases();
        const shims = this.getShimModules();
        for (const name of Object.keys(aliases))
            delete versions[name];
        for (const name of Object.keys(shims))
            delete versions[name];
        // Also remove alias targets that are already in versions (e.g. react-native-web is fetched via alias)
        if (Object.keys(versions).length === 0)
            return;
        const hash = await hashDeps(versions);
        const baseUrl = this.config.server.packageServerUrl;
        try {
            // Try GET first (CDN cacheable)
            const getRes = await fetch(`${baseUrl}/bundle-deps/${hash}`);
            if (getRes.ok) {
                const { packages } = parseDepBundle(await getRes.text());
                this.prefetchedPackages = packages;
                return;
            }
            // POST to build
            const postRes = await fetch(`${baseUrl}/bundle-deps`, {
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify({ hash, dependencies: versions }),
            });
            if (postRes.ok) {
                const { packages } = parseDepBundle(await postRes.text());
                this.prefetchedPackages = packages;
            }
        }
        catch (err) {
            // Silently fall back to individual fetches
            console.warn("[prefetch] Failed, falling back to individual fetches:", err);
        }
    }
    /** Read tsconfig.json "compilerOptions.paths" from the VirtualFS */
    static readTsconfigPaths(fs) {
        const raw = fs.read("/tsconfig.json");
        if (!raw)
            return null;
        try {
            const tsconfig = JSON.parse(raw);
            const paths = tsconfig?.compilerOptions?.paths;
            return paths && typeof paths === "object" ? paths : null;
        }
        catch {
            return null;
        }
    }
    /** Run the full pre-transform -> Sucrase -> post-transform pipeline */
    runTransform(filename, src) {
        const originalSrc = src;
        const originalLines = countNewlines(src);
        // Pre-transform hooks
        for (const plugin of this.plugins) {
            if (plugin.transformSource) {
                const result = plugin.transformSource({ src, filename });
                if (result)
                    src = result.src;
            }
        }
        const preTransformAddedLines = countNewlines(src) - originalLines;
        // Core transform (Sucrase)
        let transformResult;
        try {
            transformResult = this.config.transformer.transform({ src, filename });
        }
        catch (err) {
            throw formatTransformError(err, filename, originalSrc, src, (s) => this.config.transformer.transform({ src: s, filename }));
        }
        let code = transformResult.code;
        let sourceMap = transformResult.sourceMap;
        // Post-transform hooks -- track line additions for source map offset
        const linesBeforePost = countNewlines(code);
        for (const plugin of this.plugins) {
            if (plugin.transformOutput) {
                const result = plugin.transformOutput({ code, filename });
                if (result)
                    code = result.code;
            }
        }
        // Adjust source map for plugin modifications
        if (sourceMap) {
            // Post-transform: shift generated lines for prepended output lines
            const postAddedLines = countNewlines(code) - linesBeforePost;
            if (postAddedLines > 0) {
                sourceMap = {
                    ...sourceMap,
                    mappings: ";".repeat(postAddedLines) + sourceMap.mappings,
                };
            }
            // Pre-transform: shift origLine back so mappings point to original source
            if (preTransformAddedLines > 0) {
                sourceMap = shiftSourceMapOrigLines(sourceMap, -preTransformAddedLines);
            }
        }
        return { code, sourceMap };
    }
    /** Build a resolve callback that consults plugins then falls back to default resolution */
    makeResolveTarget(fromFile) {
        return (target) => {
            // Let plugins resolve first
            for (const plugin of this.plugins) {
                if (plugin.resolveRequest) {
                    const result = plugin.resolveRequest({ fromFile }, target);
                    if (result !== null)
                        return result;
                }
            }
            // Default resolution: skip npm packages, resolve local paths
            if (this.resolver.isNpmPackage(target))
                return null;
            const resolved = this.resolver.resolvePath(fromFile, target);
            const actual = this.resolver.resolveFile(resolved);
            return actual ?? null;
        };
    }
    /** Collect module aliases from all plugins */
    getModuleAliases() {
        const aliases = {};
        for (const plugin of this.plugins) {
            if (plugin.moduleAliases) {
                Object.assign(aliases, plugin.moduleAliases());
            }
        }
        return aliases;
    }
    /** Collect module shims from all plugins */
    getShimModules() {
        const shims = {};
        for (const plugin of this.plugins) {
            if (plugin.shimModules) {
                Object.assign(shims, plugin.shimModules());
            }
        }
        return shims;
    }
    /** Transform a single file using the configured transformer */
    transformFile(filename, src) {
        return this.runTransform(filename, src).code;
    }
    /** Bundle starting from the entry file, returning executable code */
    async bundle(entryFile) {
        const { moduleMap, sourceMapMap } = await this.buildModuleMap(entryFile);
        return this.emitBundle(moduleMap, sourceMapMap, entryFile);
    }
    /** Read dependency versions from the project's package.json */
    getPackageVersions() {
        const raw = this.fs.read("/package.json");
        if (!raw)
            return {};
        try {
            const pkg = JSON.parse(raw);
            return pkg.dependencies || {};
        }
        catch {
            return {};
        }
    }
    /** Resolve an npm specifier to a versioned form using package.json versions.
     *  Priority: user's package.json > transitive dep versions from manifests > bare name */
    resolveNpmSpecifier(specifier, versions, transitiveDepsVersions) {
        // Extract base package name: "@scope/pkg/sub" -> "@scope/pkg", "lodash/fp" -> "lodash"
        let baseName;
        if (specifier.startsWith("@")) {
            const parts = specifier.split("/");
            baseName = parts[0] + "/" + parts[1];
        }
        else {
            baseName = specifier.split("/")[0];
        }
        const version = versions[baseName] || (transitiveDepsVersions && transitiveDepsVersions[baseName]);
        if (!version)
            return specifier;
        const subpath = specifier.slice(baseName.length); // e.g. "/client" or ""
        return baseName + "@" + version + subpath;
    }
    /** Fetch a pre-bundled npm package from the package server */
    async fetchPackage(specifier) {
        // Check prefetched registry first (extract base package name from versioned specifier)
        // specifier is like "react@19.1.0" or "react-dom@19.1.0/client"
        let baseName = specifier;
        const atIdx = specifier.indexOf("@", specifier.startsWith("@") ? 1 : 0);
        if (atIdx > 0) {
            baseName = specifier.slice(0, atIdx);
            const afterVersion = specifier.indexOf("/", atIdx + 1);
            if (afterVersion > 0) {
                baseName = specifier.slice(0, atIdx) + specifier.slice(afterVersion);
            }
        }
        if (this.prefetchedPackages[baseName]) {
            return { code: lowerDynamicImports(this.prefetchedPackages[baseName]), externals: {} };
        }
        // Fallback to individual fetch
        const url = this.config.server.packageServerUrl + "/pkg/" + specifier;
        const res = await fetch(url);
        if (!res.ok) {
            const body = await res.text().catch(() => "");
            throw new Error("Failed to fetch package '" + specifier + "' (HTTP " + res.status + ")" + (body ? ": " + body.slice(0, 200) : ""));
        }
        const raw = await res.text();
        const externals = parseExternalsFromBody(raw);
        // Lower dynamic `import("x")` left behind by esbuild for externalized
        // specifiers — these otherwise hit the browser's native module loader
        // instead of our `require` registry.
        const code = lowerDynamicImports(raw);
        return { code, externals };
    }
    /** Build the module map by walking the dependency graph */
    async buildModuleMap(entryFile) {
        // Prefetch all deps in a single batch request
        await this.prefetchDependencies();
        const moduleMap = {};
        const sourceMapMap = {};
        const visited = {};
        const npmPackages = {};
        const versions = this.getPackageVersions();
        const transitiveDepsVersions = {};
        const walk = async (filePath) => {
            if (visited[filePath])
                return;
            visited[filePath] = true;
            // Asset files get a stub module that exports the filename (or a real URL for external assets)
            if (this.resolver.isAssetFile(filePath)) {
                if (this.fs.isExternalAsset(filePath) && this.config.assetPublicPath) {
                    const assetUrl = this.config.assetPublicPath + filePath;
                    moduleMap[filePath] = "module.exports = { uri: " + JSON.stringify(assetUrl) + " };";
                }
                else {
                    moduleMap[filePath] = "module.exports = " + JSON.stringify(filePath) + ";";
                }
                return;
            }
            const source = this.fs.read(filePath);
            if (source === undefined) {
                throw new Error("File not found: " + filePath);
            }
            // Transform the file (TS -> JS, JSX -> JS, etc.)
            const { code: transformed, sourceMap } = this.runTransform(filePath, source);
            if (sourceMap)
                sourceMapMap[filePath] = sourceMap;
            const rewritten = rewriteRequires(transformed, filePath, this.makeResolveTarget(filePath));
            moduleMap[filePath] = rewritten;
            const deps = findRequires(rewritten);
            for (let i = 0; i < deps.length; i++) {
                const dep = deps[i];
                if (this.resolver.isNpmPackage(dep)) {
                    if (!npmPackages[dep]) {
                        npmPackages[dep] = true;
                    }
                }
                else {
                    const resolved = this.resolver.resolveFile(dep);
                    if (resolved) {
                        await walk(resolved);
                    }
                }
            }
        };
        await walk(entryFile);
        // Process module aliases: swap sources for targets in the fetch list
        const aliases = this.getModuleAliases();
        for (const [from, to] of Object.entries(aliases)) {
            delete npmPackages[from];
            npmPackages[to] = true;
        }
        // Collect shims: these replace npm packages with inline code
        const shims = this.getShimModules();
        for (const name of Object.keys(shims)) {
            delete npmPackages[name];
        }
        // Fetch npm packages in parallel, then resolve any transitive deps (subpath requires etc.)
        const skipNames = new Set([...Object.keys(aliases), ...Object.keys(shims)]);
        const knownNpm = new Set(Object.keys(npmPackages));
        let toFetch = [...knownNpm];
        while (toFetch.length > 0) {
            const fetches = toFetch.map((name) => {
                const versionedSpecifier = this.resolveNpmSpecifier(name, versions, transitiveDepsVersions);
                return this.fetchPackage(versionedSpecifier).then(({ code, externals }) => {
                    moduleMap[name] = code;
                    // Merge externals into transitive versions (don't overwrite existing entries)
                    for (const [dep, ver] of Object.entries(externals)) {
                        if (!transitiveDepsVersions[dep]) {
                            transitiveDepsVersions[dep] = ver;
                        }
                    }
                });
            });
            await Promise.all(fetches);
            // Discover any npm deps from fetched packages not yet known
            toFetch = [];
            for (const name of knownNpm) {
                for (const dep of findRequires(moduleMap[name])) {
                    if (this.resolver.isNpmPackage(dep) && !knownNpm.has(dep) && !skipNames.has(dep)) {
                        knownNpm.add(dep);
                        toFetch.push(dep);
                    }
                }
            }
        }
        // Inject alias shim modules: require("react-native") → re-exports react-native-web
        for (const [from, to] of Object.entries(aliases)) {
            moduleMap[from] = 'module.exports = require("' + to + '");';
        }
        // Inject inline shim modules
        for (const [name, code] of Object.entries(shims)) {
            moduleMap[name] = code;
        }
        return { moduleMap, sourceMapMap };
    }
    /** Emit the final bundle string */
    emitBundle(moduleMap, sourceMapMap, entryFile) {
        const preamble = buildBundlePreamble(this.config.env, this.config.routerShim);
        const runtimeStr = "(function(modules) {\n" +
            "  var cache = {};\n" +
            "  function require(id) {\n" +
            "    if (cache[id]) return cache[id].exports;\n" +
            "    if (!modules[id]) throw new Error('Module not found: ' + id);\n" +
            "    var module = cache[id] = { exports: {} };\n" +
            "    modules[id].call(module.exports, module, module.exports, require);\n" +
            "    return module.exports;\n" +
            "  }\n" +
            "  require(" +
            JSON.stringify(entryFile) +
            ");\n" +
            "})({\n";
        const headerStr = preamble + runtimeStr;
        const ids = Object.keys(moduleMap);
        const moduleEntries = ids
            .map((id) => {
            return (JSON.stringify(id) +
                ": function(module, exports, require) {\n" +
                moduleMap[id] +
                "\n}");
        })
            .join(",\n\n");
        let bundle = headerStr + moduleEntries + "\n});\n";
        // Build combined source map
        let lineOffset = countNewlines(headerStr);
        const inputs = [];
        for (let i = 0; i < ids.length; i++) {
            const id = ids[i];
            if (i > 0)
                lineOffset += 2; // ",\n\n"
            lineOffset += 1; // wrapper function line
            if (sourceMapMap[id]) {
                const sourceContent = this.fs.read(id);
                if (sourceContent) {
                    inputs.push({
                        sourceFile: id,
                        sourceContent,
                        map: sourceMapMap[id],
                        generatedLineOffset: lineOffset,
                    });
                }
            }
            lineOffset += countNewlines(moduleMap[id]);
            lineOffset += 1; // "\n}"
        }
        if (inputs.length > 0) {
            bundle += inlineSourceMap(buildCombinedSourceMap(inputs)) + "\n";
        }
        return bundle;
    }
}
