// browser-metro/dist/fs.js
export class VirtualFS {
    constructor(files) {
        this.files = { ...files };
    }
    read(path) {
        const entry = this.files[path];
        if (entry === undefined)
            return undefined;
        return entry.content;
    }
    write(path, content) {
        this.files[path] = { content, isExternal: false };
    }
    delete(path) {
        if (path in this.files) {
            delete this.files[path];
            return true;
        }
        return false;
    }
    exists(path) {
        return path in this.files;
    }
    /** Check if a file is an external asset (binary file served from public/) */
    isExternalAsset(path) {
        const entry = this.files[path];
        return entry !== undefined && entry.isExternal === true;
    }
    list() {
        return Object.keys(this.files);
    }
    getEntryFile() {
        // Try common entry points
        const candidates = [
            "/index.js",
            "/index.ts",
            "/index.tsx",
            "/index.jsx",
            "/App.js",
            "/App.ts",
            "/App.tsx",
            "/App.jsx",
        ];
        for (const c of candidates) {
            if (this.exists(c))
                return c;
        }
        // Check package.json "main" field for local entry
        const pkgJson = this.read("/package.json");
        if (pkgJson) {
            try {
                const pkg = JSON.parse(pkgJson);
                if (typeof pkg.main === "string" && (pkg.main.startsWith(".") || pkg.main.startsWith("/"))) {
                    const base = pkg.main.replace(/^\.\//, "/").replace(/^([^/])/, "/$1");
                    if (this.exists(base))
                        return base;
                    for (const ext of [".js", ".ts", ".tsx", ".jsx"]) {
                        if (this.exists(base + ext))
                            return base + ext;
                    }
                }
            }
            catch { }
        }
        return null;
    }
    /** Read the "main" field from package.json, if present */
    getPackageMain() {
        const pkgJson = this.read("/package.json");
        if (!pkgJson)
            return null;
        try {
            const pkg = JSON.parse(pkgJson);
            return typeof pkg.main === "string" ? pkg.main : null;
        }
        catch {
            return null;
        }
    }
    toFileMap() {
        return { ...this.files };
    }
    /** Compare current state against an incoming FileMap, return list of changes */
    diff(newFiles) {
        const changes = [];
        // Check for creates and updates
        for (const path in newFiles) {
            if (!(path in this.files)) {
                changes.push({ path, type: "create" });
            }
            else if (this.files[path].content !== newFiles[path].content) {
                changes.push({ path, type: "update" });
            }
        }
        // Check for deletes
        for (const path in this.files) {
            if (!(path in newFiles)) {
                changes.push({ path, type: "delete" });
            }
        }
        return changes;
    }
    /** Replace all files at once */
    replaceAll(files) {
        this.files = { ...files };
    }
}
