// browser-metro/dist/utils.js
import { parse as acornParse } from "acorn";
const REQUIRE_RE = /require\s*\(\s*['"]([^'"]+)['"]\s*\)/g;
// Dynamic `import("x")`. Use a negative lookbehind to avoid matching `.import(`
// (e.g. method calls) and require `import` to be a standalone keyword.
const DYNAMIC_IMPORT_RE = /(?<![.$\w])import\s*\(\s*['"]([^'"]+)['"]\s*\)/g;
// Matches lines that are single-line comments or JSDoc/block comment continuations
const COMMENT_LINE_RE = /^\s*(?:\/\/|\/?\*)/;
const EXTERNALS_RE = /^\/\/ @externals (.+)$/m;
const DEP_MANIFEST_RE = /^\/\/ @dep-manifest (.+)$/m;
const DEP_START_RE = /^\/\/ @dep-start (.+)$/gm;
const DEP_END_RE = /^\/\/ @dep-end (.+)$/gm;
/** Parse externals metadata from a package bundle body.
 *  Looks for a `// @externals {...}` comment line near the top.
 *  Falls back to empty object if not found. */
export function parseExternalsFromBody(code) {
    const match = EXTERNALS_RE.exec(code);
    if (!match)
        return {};
    try {
        return JSON.parse(match[1]);
    }
    catch {
        return {};
    }
}
/** Extract all require('...') and dynamic import('...') call targets from source */
export function findRequires(source) {
    if (!source)
        return [];
    const requires = [];
    const lines = source.split("\n");
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        // Skip comment lines to avoid picking up require() in JSDoc examples
        if (COMMENT_LINE_RE.test(line))
            continue;
        const reqRe = new RegExp(REQUIRE_RE.source, REQUIRE_RE.flags);
        let match;
        while ((match = reqRe.exec(line)) !== null) {
            requires.push(match[1]);
        }
        const impRe = new RegExp(DYNAMIC_IMPORT_RE.source, DYNAMIC_IMPORT_RE.flags);
        while ((match = impRe.exec(line)) !== null) {
            requires.push(match[1]);
        }
    }
    return requires;
}
/**
 * Rewrite require calls using a resolve callback.
 * resolveTarget returns a string to rewrite to, or null to leave unchanged.
 */
export function rewriteRequires(source, fromFile, resolveTarget) {
    const requireRewritten = source.replace(REQUIRE_RE, (full, target) => {
        const resolved = resolveTarget(target);
        if (resolved === null)
            return full;
        return 'require("' + resolved + '")';
    });
    // Lower dynamic `import("x")` so it flows through the same module registry
    // as static requires. Wrap the result with __esModule interop so callers
    // like React.lazy — which expect a module namespace with `.default` —
    // work for plain CJS modules whose `module.exports` is the value itself.
    return rewriteDynamicImports(requireRewritten, (target) => {
        const resolved = resolveTarget(target);
        return resolved === null ? target : resolved;
    });
}
/**
 * Lower dynamic `import("x")` without rewriting require targets. Used for
 * prebundled npm package code where specifiers are already registry keys.
 */
export function lowerDynamicImports(source) {
    return rewriteDynamicImports(source, (target) => target);
}
function loweredDynamicImport(id) {
    return ('Promise.resolve().then(function(){var m=require("' +
        id +
        '");return m&&m.__esModule?m:{default:m};})');
}
/**
 * AST-based lowering of `import("x")` calls using acorn. Walks the parsed
 * tree and only rewrites real `ImportExpression` nodes whose argument is a
 * string literal, so source that merely *mentions* `import(...)` inside
 * strings, template literals, regex literals, or comments (e.g. React's
 * "lazy: Expected the result of a dynamic import()" error message) is
 * left untouched. Falls back to returning the source unchanged if parsing
 * fails — the regex-based passes that consume this output handle the rest.
 */
function rewriteDynamicImports(source, resolveId) {
    let ast;
    try {
        ast = acornParse(source, {
            ecmaVersion: "latest",
            sourceType: "module",
            allowImportExportEverywhere: true,
            allowAwaitOutsideFunction: true,
            allowReturnOutsideFunction: true,
            allowHashBang: true,
        });
    }
    catch {
        return source;
    }
    const edits = [];
    const visit = (node) => {
        if (!node || typeof node !== "object" || typeof node.type !== "string")
            return;
        if (node.type === "ImportExpression" &&
            node.source &&
            node.source.type === "Literal" &&
            typeof node.source.value === "string") {
            edits.push({
                start: node.start,
                end: node.end,
                target: node.source.value,
            });
            return;
        }
        for (const key of Object.keys(node)) {
            if (key === "start" || key === "end" || key === "loc" || key === "range")
                continue;
            const v = node[key];
            if (Array.isArray(v)) {
                for (const item of v)
                    visit(item);
            }
            else if (v && typeof v === "object" && typeof v.type === "string") {
                visit(v);
            }
        }
    };
    visit(ast);
    if (edits.length === 0)
        return source;
    edits.sort((a, b) => b.start - a.start);
    let out = source;
    for (const e of edits) {
        const id = resolveId(e.target);
        out = out.slice(0, e.start) + loweredDynamicImport(id) + out.slice(e.end);
    }
    return out;
}
const PUBLIC_ENV_PREFIXES = ["EXPO_PUBLIC_", "NEXT_PUBLIC_"];
/**
 * Build the bundle preamble (Metro-style). Defines `process` global so npm
 * packages that check `process.env.NODE_ENV` work in the browser.
 * Optionally injects public env vars (EXPO_PUBLIC_*, NEXT_PUBLIC_*).
 * Optionally appends the router shim for virtualizing History/Location APIs.
 */
export function buildBundlePreamble(env, routerShim) {
    let preamble = "var process = globalThis.process || {};\n" +
        "process.env = process.env || {};\n" +
        'process.env.NODE_ENV = process.env.NODE_ENV || "development";\n';
    if (env) {
        for (const [key, value] of Object.entries(env)) {
            if (PUBLIC_ENV_PREFIXES.some((p) => key.startsWith(p))) {
                preamble += "process.env." + key + " = " + JSON.stringify(value) + ";\n";
            }
        }
    }
    if (routerShim) {
        preamble += buildRouterShim();
    }
    return preamble;
}
/**
 * Build a self-contained IIFE that virtualizes the History API.
 *
 * The iframe is loaded via a blob URL with the initial route encoded in the
 * hash fragment (e.g. blob:origin/uuid#/explore). Location properties are
 * [LegacyUnforgeable] so we can't override them, but we CAN call the real
 * replaceState to change the blob URL's pathname from the UUID to the
 * virtual path. After that, location.pathname returns the correct value
 * for Expo Router's initial route match.
 *
 * All subsequent pushState/replaceState calls are intercepted so no further
 * real URL changes occur. The current virtual route is stored in
 * window.__ROUTER_SHIM_HASH__ for the parent to read across iframe rebuilds.
 */
export function buildRouterShim() {
    return `(function() {
  var rawHash = location.hash.slice(1) || '/';

  var hashIdx = rawHash.indexOf('#');
  var virtualHash = '';
  var pathAndSearch = rawHash;
  if (hashIdx > 0) {
    virtualHash = rawHash.slice(hashIdx);
    pathAndSearch = rawHash.slice(0, hashIdx);
  }

  var searchIdx = pathAndSearch.indexOf('?');
  var virtualPathname = searchIdx >= 0 ? pathAndSearch.slice(0, searchIdx) : pathAndSearch;
  var virtualSearch = searchIdx >= 0 ? pathAndSearch.slice(searchIdx) : '';

  if (virtualPathname.charAt(0) !== '/') virtualPathname = '/' + virtualPathname;

  var virtualOrigin = location.origin || 'http://localhost';
  var virtualHref = virtualOrigin + virtualPathname + virtualSearch + virtualHash;

  // Override document.URL (configurable, unlike location.*)
  try {
    Object.defineProperty(Document.prototype, 'URL', {
      get: function() { return virtualHref; },
      configurable: true
    });
  } catch(e) {}

  // Wrap URL constructor so new URL(location.href) returns the virtual URL
  // instead of parsing the blob UUID as a pathname.
  var OrigURL = URL;
  var _URL = function(url, base) {
    var u = String(url);
    if (u.indexOf('blob:') === 0) u = virtualHref;
    if (arguments.length > 1) return new OrigURL(u, base);
    return new OrigURL(u);
  };
  _URL.prototype = OrigURL.prototype;
  _URL.createObjectURL = OrigURL.createObjectURL;
  _URL.revokeObjectURL = OrigURL.revokeObjectURL;
  _URL.canParse = OrigURL.canParse;
  window.URL = _URL;

  // Save real replaceState before overriding
  var _realReplaceState = history.replaceState;
  // Base blob URL without hash, for constructing absolute URLs in sync()
  var _blobBase = location.href.split('#')[0];

  var stack = [{ state: null, pathname: virtualPathname, search: virtualSearch, hash: virtualHash }];
  var stackIndex = 0;

  function currentEntry() { return stack[stackIndex]; }

  function sync() {
    var e = currentEntry();
    virtualPathname = e.pathname;
    virtualSearch = e.search;
    virtualHash = e.hash;
    virtualHref = virtualOrigin + virtualPathname + virtualSearch + virtualHash;
    var routeHash = '#' + e.pathname + e.search;
    window.__ROUTER_SHIM_HASH__ = routeHash;
    // Update the real blob URL hash, but only if it actually changed.
    var newUrl = _blobBase + routeHash;
    if (newUrl !== location.href) {
      try { _realReplaceState.call(history, e.state, '', newUrl); } catch(e) {}
    }
  }

  function parseUrl(url) {
    var s = String(url);
    if (s.indexOf('blob:') === 0) {
      try { var inner = new OrigURL(s.slice(5)); s = inner.pathname + inner.search + inner.hash; } catch(e) {}
    }
    try {
      var u = new OrigURL(s, virtualOrigin);
      return { pathname: u.pathname, search: u.search, hash: u.hash };
    } catch(e) {}
    var pathname = s, search = '', hash = '';
    var hi = pathname.indexOf('#');
    if (hi >= 0) { hash = pathname.slice(hi); pathname = pathname.slice(0, hi); }
    var si = pathname.indexOf('?');
    if (si >= 0) { search = pathname.slice(si); pathname = pathname.slice(0, si); }
    if (!pathname) pathname = '/';
    if (pathname.charAt(0) !== '/') pathname = '/' + pathname;
    return { pathname: pathname, search: search, hash: hash };
  }

  history.pushState = function(state, title, url) {
    if (url != null) {
      var p = parseUrl(url);
      // Ignore hash: it's our own route hash echoed back from location.hash
      stack = stack.slice(0, stackIndex + 1);
      stack.push({ state: state, pathname: p.pathname, search: p.search, hash: '' });
      stackIndex = stack.length - 1;
      sync();
    }
  };

  history.replaceState = function(state, title, url) {
    if (url != null) {
      var p = parseUrl(url);
      stack[stackIndex] = { state: state, pathname: p.pathname, search: p.search, hash: '' };
      sync();
    }
  };

  history.go = function(n) {
    if (!n) return;
    var ni = stackIndex + n;
    if (ni < 0) ni = 0;
    if (ni >= stack.length) ni = stack.length - 1;
    if (ni === stackIndex) return;
    stackIndex = ni;
    sync();
    window.dispatchEvent(new PopStateEvent('popstate', { state: currentEntry().state }));
  };
  history.back = function() { history.go(-1); };
  history.forward = function() { history.go(1); };

  Object.defineProperty(history, 'state', {
    get: function() { return currentEntry().state; },
    configurable: true
  });

  sync();
})();
`;
}
/** Fast djb2 hash for cache invalidation */
export function hashString(str) {
    let hash = 5381;
    for (let i = 0; i < str.length; i++) {
        hash = ((hash << 5) + hash + str.charCodeAt(i)) | 0;
    }
    return (hash >>> 0).toString(36);
}
// Must match SERVER_VERSION in reactnative-esm/src/index.ts
const DEPS_HASH_VERSION = "2";
/** Hash a dependencies object to a stable cache key.
 *  Uses SHA-256 (via Web Crypto or Node crypto) truncated to 16 hex chars
 *  for collision resistance while keeping URLs short.
 *  Includes a version prefix so cache is invalidated when bundling logic changes. */
export async function hashDeps(deps) {
    const sorted = Object.keys(deps).sort().map(k => `${k}@${deps[k]}`).join(",");
    const input = `v${DEPS_HASH_VERSION}:${sorted}`;
    // Web Crypto API (works in browsers and workers)
    if (typeof globalThis.crypto?.subtle?.digest === "function") {
        const data = new TextEncoder().encode(input);
        const buf = await crypto.subtle.digest("SHA-256", data);
        const arr = Array.from(new Uint8Array(buf));
        return arr.map(b => b.toString(16).padStart(2, "0")).join("").slice(0, 16);
    }
    // Fallback: djb2 (Node.js without crypto, shouldn't normally happen)
    return hashString(input);
}
/** Parse a dep bundle response into individual package code chunks.
 *  Format: `// @dep-start <name>\n..code..\n// @dep-end <name>` */
export function parseDepBundle(code) {
    let manifest = {};
    const manifestMatch = DEP_MANIFEST_RE.exec(code);
    if (manifestMatch) {
        try {
            manifest = JSON.parse(manifestMatch[1]);
        }
        catch { }
    }
    const packages = {};
    const startRe = /^\/\/ @dep-start (.+)$/gm;
    let match;
    while ((match = startRe.exec(code)) !== null) {
        const name = match[1];
        const startIdx = match.index + match[0].length + 1; // skip newline
        const endMarker = `// @dep-end ${name}`;
        const endIdx = code.indexOf(endMarker, startIdx);
        if (endIdx !== -1) {
            packages[name] = code.slice(startIdx, endIdx).trimEnd();
        }
    }
    return { manifest, packages };
}
